### PR TITLE
eslint-config-standard-with-typescript

### LINKED ISSUE #22: Lint
There should be linting as part of test script.

### PR DESCRIPTION
**Pull request by mightyiam**
Fixes #22.

Well, this one is about taste.

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1,30 +0,0 @@
...
-        "quotes": [
-            "error",
-            "single"
-        ],
-        "semi": [
-            "error",
-            "always"
-        ]
-    }
-};
```
Comment (mightyiam) [inline .eslintrc.js]: `.eslintrc.json`. because I'm not sure how well the VSCode extension plays with the `.js` one.

File: .eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,4 @@
+{
+  "extends": "standard-with-typescript",
```
Comment (mightyiam) [inline .eslintrc.json]: https://eslint.org/docs/user-guide/configuring#extending-configuration-files

File: .eslintrc.json [hunk 3]
```diff
@@ -0,0 +1,4 @@
...
+  "parser": "typescript-eslint-parser"
```
Comment (mightyiam) [inline .eslintrc.json]: https://github.com/eslint/typescript-eslint-parser/

File: .vscode/settings.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+  "eslint.validate": [
+    "typescript"
```
Comment (mightyiam) [inline .vscode/settings.json]: This is to make VS Code users happy. Their ESLint extension will lint the source files.

File: .vscode/settings.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+  ],
+  "tslint.enable": false
```
Comment (mightyiam) [inline .vscode/settings.json]: And tell TSLint to stay out of it.

File: package.json [hunk 6]
```diff
@@ -16,9 +16,10 @@
     "node": ">=4"
   },
   "scripts": {
+    "lint": "eslint --ext .ts src/**/*.ts",
```
Comment (mightyiam) [inline package.json]: https://eslint.org/docs/user-guide/configuring#specifying-file-extensions-to-lint

File: package.json [hunk 7]
```diff
@@ -37,10 +38,17 @@
   "devDependencies": {
     "ava": "^0.23.0",
     "coveralls": "^3.0.0",
+    "eslint": "^4.10.0",
+    "eslint-config-standard-with-typescript": "^1.0.6",
```
Comment (mightyiam) [inline package.json]: https://www.npmjs.com/package/eslint-config-standard-with-typescript

File: src/index.test.ts [hunk 8]
```diff
@@ -1,127 +1,127 @@
...
+import { spy } from './'
 
 test('should call the original function', t => {
-    // Arrange
-    let called = false;
-    const fun = () => called = true;
-    const funSpy = spy(fun);
+  // Arrange
+  let called = false
+  const fun = () => { called = true }
```
Comment (mightyiam) [inline src/index.test.ts]: https://eslint.org/docs/rules/no-return-assign

File: src/index.test.ts [hunk 9]
```diff
@@ -1,127 +1,127 @@
...
+})
 
 test('should call the original function with arguments', t => {
-    // Arrange
-    let myNumber = 0;
-    const fun = (newNumber: number) => myNumber = newNumber;
-    const funSpy = spy(fun);
+  // Arrange
+  let myNumber = 0
+  const fun = (newNumber: number) => { myNumber = newNumber }
```
Comment (mightyiam) [inline src/index.test.ts]: https://eslint.org/docs/rules/no-return-assign

File: package.json [hunk 10]
```diff
@@ -37,10 +38,17 @@
...
+    "eslint-plugin-import": "^2.8.0",
+    "eslint-plugin-node": "^5.2.1",
+    "eslint-plugin-promise": "^3.6.0",
+    "eslint-plugin-standard": "^3.0.1",
```
[the remark this answers is not in the document]
Comment (mightyiam) [reply package.json]: Sorry, [long story](https://github.com/standard/eslint-config-standard/pull/91).

**Comment by mightyiam** [comment]
Regarding the warnings: linting TypeScript with ESLint is rather new. There are known issues with typescript-eslint-parser. They are because of added node types on the AST. So, some core ESLint rules don't know to mark certain names as used. And That `no-undef` thing, as well. So, WIP. I'm staying current with it in my project, eslint-config-standard-with-typescript, which I'm using in several repos. Yeah, it is not a bargain, but at least someone is looking after it. And it is easy to swap out for something else if you don't like it. And, hey, at least one of us likes it.

**Comment by mightyiam** [comment]
https://github.com/eslint/typescript-eslint-parser/#known-issues

File: package.json (hunk 10 above)
[the remark this answers is not in the document]
Comment (tjoskar) [reply package.json]: ✅

**Comment by tjoskar** [comment]
@mightyiam, I think it is better to use [tslint](https://palantir.github.io/tslint/) if eslint don't support typescript fully yet. I don't see any reason to use eslint over tslint if tslint has better support for typescript than eslint. I do not like to have warnings in the code that we can't get rid of.
We can use this lib to get the standard js rules: https://github.com/blakeembrey/tslint-config-standard

**Comment by mightyiam** [comment]
I'm familiar with tslint-config-standard. I tried it before I made eslint-config-standard-with-typescript.

A reason to use any incomplete solution might be that we conceptually prefer it and hope that our participation will lead to a contribution towards its completion.

So, I prefer the concept of using the existing ESLint ecosystem of rules, configs and plugins with TypeScript. And for my projects I go the extra mile and endure those warnings in my IDE and even an occasional new bug. 2 or three bugs that I have reported to typescript-eslint-parser have already been fixed. I enjoy that, as well.

No pressure at all. Just my explanation. Your lib—your call.

**Comment by mightyiam** [comment]
There was progess in typescript-eslint-parser and eslint-plugin-typescript. And also my own eslint-config-standard-with-typescript. So, after #29 is merged, there should be no ESLint warnings with this PR.

**Comment by tjoskar** [comment]
`tslint` can get information from the typescript compiler and can therefor do some more advance linting. Like: [no-unnecessary-type-assertion](https://palantir.github.io/tslint/rules/no-unnecessary-type-assertion/), [await-promise](https://palantir.github.io/tslint/rules/await-promise/), [no-floating-promises](https://palantir.github.io/tslint/rules/no-floating-promises/) and many more. I think that `eslint` is a better choice in general but since this is a small project I do not have any strong opinion. I will wait with this PR untill the other ones are merged.

**Comment by tjoskar** [comment]
@mightyiam, you are right. After upgrading typescript-eslint-parser the warnings are gone. I will accept this PR. Do you want to rebase master and update the dependencies? Or should I?

**Comment by mightyiam** [comment]
Rebased.

**Comment by tjoskar** [comment]
Thanks @mightyiam

**Pull request merged**